_posts/2016-05-11-android-testing-with-kotlin.md

```diff
@@ -0,0 +1,422 @@
+
+        var items = arrayOf(
+            Pair("    item 1", "item 1"),
+            Pair("item 2  ", "item 2"),
+            Pair("  item  \n3  ", "item 3")
+        )
+
+        items.map { pair -> addItemToList(pair.first) }
+        items.mapIndexed { index, pair ->
+            val expected = formatter.strip(pair.second)
+            lv.atPosition(index).onChildView(withId(R.id.tv_list_item)).check(
+                matches(withText(expected))
+            )
+        }
+    }
+}
+```
+
+You can see that the Espresso test suite at hand doesn't differ much from the
+unit tests we discussed in the previous section. And neither does it differ much from
+your already existing Espresso tests written in Java. Truth is, that majority of the
+code that does the heavy lifting is plain *Java* and other parts are almost
+drop in replacements.
+
+As usual with instrumentation tests, we start off by creating an instrumented JUnit 4
+test class by adding
+
+```ruby
+@RunWith(AndroidJUnit4::class)
+```
+
+annotation at the beginning of our test class definition. And naturally we can make
+use of Espresso's built-in `ActivityTestRule` to handle our activity life cycle:
+
+```ruby
+@Rule @JvmField
+var activityRule = ActivityTestRule<MainActivity>(MainActivity::class.java)
+```
+
+With one minor caveat that you need to annotate it with `@JvmField` to convert it
+from Kotlin a *property*, that is unrecognized by JUnit, to a JVM field that is edible
+by JUnit.
+
+Apart from these tiny details, you can just use the familiar Espresso API for the
+test functions.
+
+## Running our Kotlin tests
+
+Since we have already updated our `sourceSets` in `build.gradle` we don't have to
+go through any configuration magic any more. You can execute both unit and instrumentation
+tests right from Android Studio by just right-clicking either on your tests packages or
+Kotlin test files and clicking on "Run 'Tests in ...'" as can be seen below:
+
+<div style="text-align:center; padding: 20px 0;"><img src="/assets/run-kotlin-tests-form-android-studio.png" alt="Run Kotlin tests from Android Studio"/></div>
+
+The very same minimal configuration logic applies also in case you should want to
+execute your tests from command line. Kotlin unit and instrumentation tests can be
+executed with regular Gradle commands
+
+```bash
+gradlew test
+```
+
+and
+
+```bash
+gradle connectedAndroidTest
+```
+
+## But what good are tests if you don't run them?
+
+Long story short, running your tests manually is just not feasible. __To ensure the
+wellbeing of your app, it has to be tested after every commit, with no exceptions allowed!__
+The best way to achieve it is to make use of CI. And with Greenhouse you get full support
+for tests written in Kotlin out of the box. Just follow our regular flow for [building
+an Android app](http://docs.greenhouseci.com/docs/building-android-apps) and we'll
+automatically detect and run all your tests, including those written in Kotlin.
+
+## Final remarks
```

Comment:
Initially I wanted to write about some pitfalls that Kotlin comes with in this last section, but decided not to. In the end I just forgot to delete the heading.